### PR TITLE
feat(hotkey): capture data part3 - declare fine collector 

### PR DESCRIPTION
**Pull request by Smityz**
the implement of fine collector 
## Config changes
```diff
[pegasus.server]
+ fine_data_variance_threshold = 3
```

### CODE REVIEW

File: src/server/hotkey_collector.cpp [hunk 1]
```diff
@@ -251,5 +261,101 @@ void hotkey_coarse_data_collector::analyse_data(detect_hotkey_result &result)
...
+    : internal_collector_base(base),
+      _hotkey_type(hotkey_type),
+      _max_queue_size(max_queue_size),
+      _target_bucket_index(target_bucket_index)
+{
+    // Distinguish between single-threaded and multi-threaded environments
+    if (_hotkey_type == dsn::replication::hotkey_type::READ) {
+        auto threads = dsn::get_threadpool_threads_info(THREAD_POOL_LOCAL_APP);
+        int queue_num = threads.size();
+        for (int i = 0; i < queue_num; i++) {
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:278]:
```
        _string_capture_queue_vec.reserve(queue_num);
        for (int i = 0; i < queue_num; i++) {
            _thread_queue_map.insert(std::make_pair(threads[i]->native_tid(), i));

            // Create a vector of the ReaderWriterQueue whose size = _max_queue_size
            _string_capture_queue_vec.emplace_back(_max_queue_size);
        }
```

File: src/server/hotkey_collector.cpp [hunk 2]
```diff
@@ -251,5 +261,101 @@ void hotkey_coarse_data_collector::analyse_data(detect_hotkey_result &result)
...
+        return dsn::string_view(lhs) == dsn::string_view(rhs);
+    }
+};
+
+int hotkey_fine_data_collector::get_queue_index()
+{
+    if (_hotkey_type == dsn::replication::hotkey_type::WRITE) {
+        return 0;
+    }
+    int thread_native_tid = ::dsn::utils::get_current_tid();
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:320]: `dsn::utils::get_current_tid()`

File: src/server/hotkey_collector.h [hunk 3]
```diff
@@ -123,5 +126,30 @@ class hotkey_coarse_data_collector : public internal_collector_base
...
+
+private:
+    inline int get_queue_index();
+
+    const dsn::replication::hotkey_type::type _hotkey_type;
+    int _max_queue_size;
+    const int _target_bucket_index;
+    // thread's native id -> data queue id.
+    std::unordered_map<int, int> _thread_queue_map;
+    string_capture_queue_vec _string_capture_queue_vec;
```
Comment (hycdong) [inline src/server/hotkey_collector.h:151]: I suggest adding some comments for this structure.

File: src/server/hotkey_collector.cpp [hunk 4]
```diff
@@ -251,5 +261,107 @@ void hotkey_coarse_data_collector::analyse_data(detect_hotkey_result &result)
...
+{
+    std::unordered_map<dsn::blob, uint64_t, blob_hash, blob_equal> hash_key_accessed_cnt;
+    for (auto &rw_queue : _string_capture_queue_vec) {
+        std::pair<dsn::blob, int> hash_key_pair;
+        // prevent endless loop
+        int collect_sum = 0;
+        while (rw_queue.try_dequeue(hash_key_pair) && ++collect_sum <= _max_queue_size) {
+            hash_key_accessed_cnt[hash_key_pair.first] += hash_key_pair.second;
+        }
+    }
```
Comment (hycdong) [inline src/server/hotkey_collector.cpp:339]: I just don't understand how to calculate haskkey_access_count.

File: src/server/hotkey_collector.cpp [hunk 5]
```diff
@@ -251,5 +261,107 @@ void hotkey_coarse_data_collector::analyse_data(detect_hotkey_result &result)
...
+
+    // if the accessed counts differ hugely (depends on the variance threshold),
+    // the max key is the hotkey.
+    int hot_index;
+    if (counts.size() < 3 ||
+        find_outlier_index(counts, FLAGS_fine_data_variance_threshold, hot_index)) {
+        result.hot_hash_key = std::string(count_max_key);
+    }
```
Comment (hycdong) [inline src/server/hotkey_collector.cpp:363]: What is the code from L357-L363 doing? In my view, the hot_hash_key is `count_max_key`, which is be found in L353.

File: src/server/hotkey_collector.cpp (hunk 4 above)
Comment (Smityz) [reply src/server/hotkey_collector.cpp:339]: https://github.com/apache/incubator-pegasus/pull/603#issue-487947949
The data will be put into different queues according to the thread, and then they will be timely export to a hashmap to calculate the frequency of occurrence

File: src/server/hotkey_collector.cpp (hunk 5 above)
Comment (Smityz) [reply src/server/hotkey_collector.cpp:363]: Only depends on the hottest data is not accurate, we must ensure that it's an outlier in its hash bucket

File: src/server/hotkey_collector.cpp [hunk 6]
```diff
@@ -251,5 +261,111 @@ void hotkey_coarse_data_collector::analyse_data(detect_hotkey_result &result)
...
+        int queue_num = threads.size();
+
+        _string_capture_queue_vec.reserve(queue_num);
+        for (int i = 0; i < queue_num; i++) {
+            _thread_queue_map.insert(std::make_pair(threads[i]->native_tid(), i));
+
+            // Create a vector of the ReaderWriterQueue whose size = _max_queue_size
+            _string_capture_queue_vec.emplace_back(_max_queue_size);
+        }
+
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:287]: remove this line

File: src/server/hotkey_collector.h [hunk 7]
```diff
@@ -123,5 +126,31 @@ class hotkey_coarse_data_collector : public internal_collector_base
...
+                                        int target_bucket_index,
+                                        int max_queue_size);
+    void capture_data(const dsn::blob &hash_key, uint64_t weight) override;
+    void analyse_data(detect_hotkey_result &result) override;
+
+private:
+    inline int get_queue_index();
+
+    const dsn::replication::hotkey_type::type _hotkey_type;
+    int _max_queue_size;
```
Comment (levy5307) [inline src/server/hotkey_collector.h:147]: use uint32_t or uint64_t or uint16_t, ...

File: src/server/hotkey_collector.h [hunk 8]
```diff
@@ -123,5 +126,31 @@ class hotkey_coarse_data_collector : public internal_collector_base
...
+    const int _target_bucket_index;
+    // thread's native id -> data queue id.
+    std::unordered_map<int, int> _thread_queue_map;
```
Comment (acelyc111) [inline src/server/hotkey_collector.h:150]: How about rename to `thread_id_to_queue_index` ?

File: src/server/hotkey_collector.h [hunk 9]
```diff
@@ -123,5 +126,31 @@ class hotkey_coarse_data_collector : public internal_collector_base
     std::vector<std::atomic<uint64_t>> _hash_buckets;
 };
 
+typedef std::vector<moodycamel::ReaderWriterQueue<std::pair<dsn::blob, uint64_t>>>
+    string_capture_queue_vec;
```
Comment (acelyc111) [inline src/server/hotkey_collector.h:130]: How about rename to `hash_key_request_times`, or some like that?

File: src/server/hotkey_collector.h [hunk 10]
```diff
@@ -123,5 +126,31 @@ class hotkey_coarse_data_collector : public internal_collector_base
...
+private:
+    inline int get_queue_index();
+
+    const dsn::replication::hotkey_type::type _hotkey_type;
+    int _max_queue_size;
+    const int _target_bucket_index;
+    // thread's native id -> data queue id.
+    std::unordered_map<int, int> _thread_queue_map;
+    // Each element in the vector corresponds to a thread, each element is a lock-free queue
+    string_capture_queue_vec _string_capture_queue_vec;
```
Comment (acelyc111) [inline src/server/hotkey_collector.h:152]: Better to give a more meanful name.

Comment (acelyc111) [inline src/server/hotkey_collector.h:152]: Better to give a more meanful name.

File: src/server/hotkey_collector.cpp [hunk 11]
```diff
@@ -251,5 +261,111 @@ void hotkey_coarse_data_collector::analyse_data(detect_hotkey_result &result)
...
+        return boost::hash_range(cp.begin(), cp.end());
+    }
+};
+struct blob_equal
+{
+    std::size_t operator()(const dsn::blob &lhs, const dsn::blob &rhs) const
+    {
+        return dsn::string_view(lhs) == dsn::string_view(rhs);
+    }
+};
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:315]: Better to move to class `dsn::blob`'s definition, then you can reuse it easily.

Comment (Smityz) [reply src/server/hotkey_collector.cpp:315]:
```cpp
    // blob.h
    bool operator==(const blob &r) const
    {
        // not implemented
        assert(false);
        return false;
    }
```
The default method is deleted, I think in different scenarios `==` has different meanings

File: src/server/hotkey_collector.cpp [hunk 12]
```diff
@@ -21,21 +21,31 @@
...
 #include <dsn/dist/fmt_logging.h>
 #include <dsn/utility/flags.h>
+#include <dsn/tool-api/task.h>
+#include <dsn/dist/replication/replication.codes.h>
+#include "base/pegasus_key_schema.h"
 
 namespace pegasus {
 namespace server {
 
 DSN_DEFINE_int32("pegasus.server",
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:33]: `DSN_DEFINE_uint32`

File: src/server/hotkey_collector.cpp [hunk 13]
```diff
@@ -21,21 +21,31 @@
...
 
 DSN_DEFINE_validator(coarse_data_variance_threshold,
                      [](int32_t threshold) -> bool { return (threshold >= 0); });
 
+DSN_DEFINE_int32("pegasus.server",
+                 fine_data_variance_threshold,
+                 3,
+                 "the threshold of variance calculate to find the outliers in fine analyse");
+
+DSN_DEFINE_validator(fine_data_variance_threshold,
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:46]: so you don't need to add `DSN_DEFINE_validator` here

File: src/server/hotkey_collector.cpp [hunk 14]
```diff
@@ -21,21 +21,31 @@
...
 DSN_DEFINE_int32("pegasus.server",
                  coarse_data_variance_threshold,
                  3,
-                 "the threshold of variance calculate to find the outliers");
+                 "the threshold of variance calculate to find the outliers in coarse analyse");
 
 DSN_DEFINE_validator(coarse_data_variance_threshold,
                      [](int32_t threshold) -> bool { return (threshold >= 0); });
 
+DSN_DEFINE_int32("pegasus.server",
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:41]: same

File: src/server/hotkey_collector.cpp [hunk 15]
```diff
@@ -251,5 +260,78 @@ void hotkey_coarse_data_collector::analyse_data(detect_hotkey_result &result)
...
+}
+
+struct blob_hash
+{
+    std::size_t operator()(const dsn::blob &str) const
+    {
+        dsn::string_view cp(str);
+        return boost::hash_range(cp.begin(), cp.end());
+    }
+};
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:288]: add a blank line

File: src/server/hotkey_collector.cpp [hunk 16]
```diff
@@ -251,5 +260,78 @@ void hotkey_coarse_data_collector::analyse_data(detect_hotkey_result &result)
...
+        return dsn::string_view(lhs) == dsn::string_view(rhs);
+    }
+};
+
+void hotkey_fine_data_collector::analyse_data(detect_hotkey_result &result)
+{
+    std::unordered_map<dsn::blob, uint64_t, blob_hash, blob_equal> hash_key_accessed_cnt;
+    std::pair<dsn::blob, int> hash_key_pair;
+    // prevent endless loop, limit the number of elements analyzed not to exceed the queue size
+    int collect_sum = 0;
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:302]: uint32_t

File: src/server/hotkey_collector.cpp (hunk 11 above)
Comment (acelyc111) [reply src/server/hotkey_collector.cpp]: How the `==` operator of a class can have different behaviors?

File: src/server/hotkey_collector.cpp [hunk 17]
```diff
@@ -251,5 +254,79 @@ void hotkey_coarse_data_collector::analyse_data(detect_hotkey_result &result)
...
+    std::size_t operator()(const dsn::blob &lhs, const dsn::blob &rhs) const
+    {
+        return dsn::string_view(lhs) == dsn::string_view(rhs);
+    }
+};
+
+void hotkey_fine_data_collector::analyse_data(detect_hotkey_result &result)
+{
+    std::unordered_map<dsn::blob, uint64_t, blob_hash, blob_equal> hash_key_accessed_cnt;
+    std::pair<dsn::blob, int> hash_key_pair;
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:295]:
```suggestion
    std::pair<dsn::blob, uint64_t> hash_key_pair;
```

Comment (acelyc111) [inline src/server/hotkey_collector.cpp:295]: `hash_key_pair ` is not good naming, what's in the "pair" except "hash_key"?

File: src/server/hotkey_collector.cpp [hunk 18]
```diff
@@ -251,5 +254,79 @@ void hotkey_coarse_data_collector::analyse_data(detect_hotkey_result &result)
...
+    // prevent endless loop, limit the number of elements analyzed not to exceed the queue size
+    uint32_t collect_sum = 0;
```
Comment (acelyc111) [inline src/server/hotkey_collector.cpp:297]: It's a 'counter' not a 'sum'...

File: src/server/hotkey_collector.h [hunk 19]
```diff
@@ -115,6 +117,7 @@ class hotkey_empty_data_collector : public internal_collector_base
 class hotkey_coarse_data_collector : public internal_collector_base
 {
 public:
+    hotkey_coarse_data_collector() = delete;
```
Comment (neverchanje) [inline src/server/hotkey_collector.h:120]: Move deleted constructor to private.

File: src/server/hotkey_collector.cpp [hunk 20]
```diff
@@ -21,26 +21,30 @@
...
 namespace server {
 
-DSN_DEFINE_int32("pegasus.server",
-                 coarse_data_variance_threshold,
-                 3,
-                 "the threshold of variance calculate to find the outliers");
+DSN_DEFINE_uint32("pegasus.server",
+                  coarse_data_variance_threshold,
+                  3,
+                  "the threshold of variance calculate to find the outliers in coarse analyse");
```
Comment (neverchanje) [inline src/server/hotkey_collector.cpp:36]:
```suggestion
DSN_DEFINE_uint32("pegasus.server",
                  hot_bucket_variance_threshold,
                  3,
                  "the variance threshold to detect hot bucket during coarse analysis of hotkey detection");
```

File: src/server/hotkey_collector.cpp [hunk 21]
```diff
@@ -21,26 +21,30 @@
...
 
-DSN_DEFINE_validator(coarse_data_variance_threshold,
-                     [](int32_t threshold) -> bool { return (threshold >= 0); });
+DSN_DEFINE_uint32("pegasus.server",
+                  fine_data_variance_threshold,
+                  3,
+                  "the threshold of variance calculate to find the outliers in fine analyse");
```
Comment (neverchanje) [inline src/server/hotkey_collector.cpp:41]:
```suggestion
DSN_DEFINE_uint32("pegasus.server",
                  hot_key_variance_threshold,
                  3,
                  "the variance threshold to detect hot key during fine analysis of hotkey detection");
```

**Review by acelyc111 — APPROVED** [review]

File: src/server/hotkey_collector.cpp [hunk 22]
```diff
@@ -21,26 +21,30 @@
...
 
 // TODO: (Tangyanzhao) add a limit to avoid changing when detecting
-DSN_DEFINE_int32("pegasus.server",
-                 data_capture_hash_bucket_num,
-                 37,
-                 "the number of data capture hash buckets");
+DSN_DEFINE_uint32("pegasus.server",
+                  data_capture_hash_bucket_num,
```
Comment (neverchanje) [inline src/server/hotkey_collector.cpp:45]:
```suggestion
                  hotkey_buckets_num,
```

File: src/server/hotkey_collector.h [hunk 23]
```diff
@@ -123,5 +126,22 @@ class hotkey_coarse_data_collector : public internal_collector_base
     std::vector<std::atomic<uint64_t>> _hash_buckets;
 };
 
+class hotkey_fine_data_collector : public internal_collector_base
+{
+public:
+    hotkey_fine_data_collector() = delete;
```
Comment (neverchanje) [inline src/server/hotkey_collector.h:132]:
```suggestion
    hotkey_fine_data_collector() = delete;
```

move deleted to private

File: src/server/hotkey_collector.cpp (hunk 21 above)
Comment (Smityz) [reply src/server/hotkey_collector.cpp:41]: Why combine two threshold

File: src/server/hotkey_collector.h [hunk 24]
```diff
@@ -123,5 +126,22 @@ class hotkey_coarse_data_collector : public internal_collector_base
...
+    explicit hotkey_fine_data_collector(replica_base *base,
+                                        int target_bucket_index,
+                                        int max_queue_size);
+    void capture_data(const dsn::blob &hash_key, uint64_t weight) override;
+    void analyse_data(detect_hotkey_result &result) override;
+
+private:
+    uint32_t _max_queue_size;
```
Comment (neverchanje) [inline src/server/hotkey_collector.h:140]: Will `_max_queue_size` change? Can it be const?

**Review by hycdong — APPROVED** [review]

**Review by neverchanje — APPROVED** [review]

**Pull request merged**